### ISSUE TITLE
Q ExponentiatedGradient results vary using the same model and data

### ISSUE THREAD

**Issue by CarlaFernandez** [fairlearn/fairlearn#1261]
### Problem

Hello, 
I'm using the ExponentiatedGradient to perform mitigations on a Catboost classifier. Each time I run the `fit` and obtain the MetricsFrame for my sensitive features, I get different results. I would like to know if this is normal, or if I may be missing some `seed` parameter somewhere. 

### Reproducible example
I'm working with the census data, and this is my code:
```python
from catboost import CatBoostClassifier
from fairlearn.metrics import MetricFrame
from fairlearn.reductions import ExponentiatedGradient, DemographicParity
from functools import partial
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.datasets import fetch_openml
from sklearn.metrics import accuracy_score, recall_score, precision_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
import fairlearn
import pandas as pd

data = fetch_openml(data_id=1590, as_frame=True)
X = data.data
y = LabelEncoder().fit_transform(data.target) 

params = {
    "iterations": 100,
    "learning_rate": 0.1,
    "depth": 7,
    "loss_function": 'Logloss',
    "eval_metric": 'Accuracy',
    "verbose": False,
    "random_state": 42
}

categorical_features = list(X.select_dtypes(include=['category', 'object']).columns)
numeric_features = list(X.select_dtypes(include=['int64', 'float64']).columns)
for feat in categorical_features:
    X[feat] = X[feat].astype(str)
    
X.loc[:, categorical_features] = X.loc[:, categorical_features].fillna("Nulo")

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

class CatBoostClassifierAdapter(BaseEstimator, ClassifierMixin):
    def __init__(self, cat_features=None, text_features=None, **kwargs):
        self.cat_features = cat_features
        self.text_features = text_features
        if "sample_weight" in kwargs.keys():
            kwargs["class_weights"] = kwargs["sample_weight"]
            kwargs.pop("sample_weight")
        self.model = CatBoostClassifier(**kwargs)
        
    def fit(self, X, y, **kwargs):
        self.model.fit(
            X, y, 
            cat_features=self.cat_features, 
            text_features=self.text_features,
            verbose=False, **kwargs
        )
        return self.model
    
    def predict(self, X):
        check_is_fitted(self.model)
        return self.model.predict(X)

model = CatBoostClassifierAdapter(cat_features=categorical_features, **params)

expgrad = ExponentiatedGradient(
    model,
    constraints=DemographicParity(),
    max_iter=5
)

expgrad.fit(X_train, y_train, sensitive_features=X_train[['sex', 'race']])

predictions = expgrad.predict(X_test)

accuracy = accuracy_score(y_test, predictions)
print(f"Accuracy: {accuracy}")

sensitive_features_test = X_test[['sex', 'race']]

recall_func = partial(recall_score, zero_division=0.0)
recall_func.__name__ = "recall"

precision_func = partial(precision_score, zero_division=0.0)
precision_func.__name__ = "precision"

metrics = MetricFrame(metrics={"recall": recall_func, "count": fairlearn.metrics.count},
                      y_true=y_test,
                      y_pred=predictions,
                      sensitive_features=sensitive_features_test)

by_group_df = metrics.by_group.reset_index()
by_group_df
```

**Comment by romanlutz** [comment]
Yes, you're missing the `random_state` parameter on `predict`:

![image](https://github.com/fairlearn/fairlearn/assets/10245648/d82a39f9-e972-4f8c-b859-6ad5aa8b9b10)

https://fairlearn.org/v0.8/api_reference/fairlearn.reductions.html#fairlearn.reductions.ExponentiatedGradient.predict


Please reopen if this doesn't fully address your question.

**Comment by CarlaFernandez** [comment]
Hello @romanlutz, indeed by setting the random_state parameter I can obtain consistent predictions using the same fitted instance of ExponentiatedGradient. However, if I refit the ExponentiatedGradient I lose reproducibility again. I can partly solve it by setting a numpy seed before fitting, and that way the results are reproducible in a single system, even restarting my jupyter kernel, but not if I move for example from my windows to linux PCs. 

I was especially interested in this kind of reproducibility because I'm writing a technical post and I would have liked the readers to be able to reexecute the notebook and obtain exactly the same results. But I guess it's just due to the nature of the algorithm that it's not possible.

Thanks anyways!

**Comment by romanlutz** [comment]
@MiroDudik this sounds like a question for you.

**Comment by MiroDudik** [comment]
@CarlaFernandez -- there shouldn't be any randomness in the exponentiated gradient during fitting. we tried hard to only use deterministic optimizers. my guess is that what you see is somehow coming from CatBoostClassifier and/or possibly from how we clone the classifiers internally here: https://github.com/fairlearn/fairlearn/blob/cb9571ea877dcd7967244760cb7a642cacfe95da/fairlearn/reductions/_exponentiated_gradient/_lagrangian.py#L227

So, one idea would be to compare whether the following three classifiers are returning the same model:

```python
from sklearn import clone

# your code setting up CatBoostClassifierAdapter and params

model1 = CatBoostClassifierAdapter(cat_features=categorical_features, **params)
model2 = CatBoostClassifierAdapter(cat_features=categorical_features, **params)
model3 = clone(estimator=model1, safe=False)

model1.fit(...)
model2.fit(...)
model3.fit(...)
```

**Comment by CarlaFernandez** [comment]
Hello @MiroDudik , how would I go about checking that model1, model2 and model3 are the same model? I can see that all of them return the same predictions, is that enough to ensure they are the same model? (I would think so)

**Comment by hildeweerts** [comment]
Hi @CarlaFernandez, I am not super familiar with the `catboost` library, but you could inspect the calculated  [feature importances](https://catboost.ai/docs/concepts/python-reference_catboostclassifier_attributes) or [plot the tree](https://catboost.ai/en/docs/concepts/python-reference_catboostclassifier_plot_tree) to check whether the models are the same.

**Comment by CarlaFernandez** [comment]
Thanks Hilde, good idea! Doing that, I see that the model importances change after cloning:
![image](https://github.com/fairlearn/fairlearn/assets/9136877/81cbd9ef-ce22-4b80-91e4-12fd83c0943c)

**Comment by MiroDudik** [comment]
It seems that `sklearn.clone` doesn't work as expected. Can you also compare with the following model?

```python
import copy

model4 = copy.deepcopy(model1)
```

Thanks so much for helping us chase this bug!

**Comment by CarlaFernandez** [comment]
Using deepcopy the model gets copied correctly!
![image](https://github.com/fairlearn/fairlearn/assets/9136877/ca42b285-ce8e-4aba-a55e-82380d9e8b4d)

**Comment by MiroDudik** [comment]
We have a bugfix. We'll just need to come up with a small regression test that would not create a dependence on catboost...

**Comment by hildeweerts** [comment]
I suspect that there is some kind of bug in the [`CatBoostClassifier.get_params()`](https://github.com/catboost/catboost/blob/02f13d9a63528757947caa05db299183b8bd2238/catboost/python-package/catboost/core.py#L3351C35-L3351C35) related to the `random_state` parameter, which is used as a synonym for the `random_seed` parameter. 

For some reason i have some issues with installing catboost, @CarlaFernandez could you let me know what you get when you do:
```python
model1.get_params()
```

You get a very similar result if you delete the `random_state` of `get_params()` of a `RandomForestClassifier`:

```python
import copy
from sklearn import clone
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier

data=load_iris()
X = data.data
y = data.target

class RandomForestClassifierTest(RandomForestClassifier):
    def get_params(self,**kwargs):
        params = super().get_params(**kwargs)
        del params['random_state']
        return params
    
model1 = RandomForestClassifierTest(random_state=0)
model2 = RandomForestClassifierTest(random_state=0)
model3 = clone(estimator=model1, safe=False)
model4 = copy.deepcopy(model1)

print(model1.fit(X,y).feature_importances_)
print(model2.fit(X,y).feature_importances_)
print(model3.fit(X,y).feature_importances_)
print(model4.fit(X,y).feature_importances_)
```

which outputs:

```
[0.09090795 0.02453104 0.46044474 0.42411627]
[0.09090795 0.02453104 0.46044474 0.42411627]
[0.10589901 0.02723903 0.42592277 0.44093919]
[0.09090795 0.02453104 0.46044474 0.42411627]
```

**Comment by CarlaFernandez** [comment]
Hi @hildeweerts , this is the output I obtain from `model1.get_params()`:
```
{'cat_features': ['workclass',
  'education',
  'marital-status',
  'occupation',
  'relationship',
  'race',
  'sex',
  'native-country'],
 'text_features': None}
```
I guess what you wanted to see are the hyperparameters for the actual catboost model, not the adapter, so here they are (`model1.model.get_params()`):
```
{'iterations': 100,
 'learning_rate': 0.1,
 'depth': 7,
 'loss_function': 'Logloss',
 'verbose': False,
 'eval_metric': 'Accuracy',
 'random_state': 42}
```
`model1`, `model2` and `model4` all have those same params. `model3` returns an empty dictionary, but I think that's expected giving the sklearn clone functionality.

I'm attaching my requirements.txt file in case it helps you with the catboost installation or debugging the problem:
[requirements.zip](https://github.com/fairlearn/fairlearn/files/12250925/requirements.zip)

Thank you :)

**Comment by MiroDudik** [comment]
@CarlaFernandez -- together with @romanlutz and @hildeweerts , we played a bit with your implementation, and the issue is in how `get_params` is implemented by `BaseEstimator`--and inhereted by `CatBoostClassifierAdapter`. More precisely, `get_params` only returns the *keyword parameters from the construtor* (not kwargs) and only those parameters get cloned (with some caveats) by `clone`. This means that the only two parameters that get cloned in `CatBoostClassifierAdapter` are `cat_features` and `text_features`. If you want to work with the current `fairlearn` code, you might want to try to refactor `CatBoostClassifierAdapter` as follows:
1. replace kwargs in the `__init__` with a dictionary argument (so you would directly pass `params` rather than `**params` to `CatBoostClassifierAdapter`)
2. in `__init__` only save the arguments into the fields, and postpone any parameter processing (including construction of `CatBoostClassifier`) to `fit`

That said, I still think that we should still replace `clone` by `deepcopy` (for the reason I mention in my previous comment). Are there any reasons why we shouldn't do it? [tagging @adrinjalali]

**Comment by romanlutz** [comment]
I agree with @MiroDudik on replacing clone with deepcopy, but curious what @adrinjalali thinks as well.

Adding a bit to what @MiroDudik wrote regarding the code. Perhaps we can resolve this for your particular problem @CarlaFernandez, too, so that you don't have to wait for a release. 

Can you elaborate on the purpose of the adapter class?
It seems like a `sample_weight` arg in the constructor is remapped to `class_weights`. Usually, `sample_weight` is passed in `fit`, though. Also, it doesn't seem like it's actually used. For the code below I've removed that part.
I also moved some things around and got it to work just fine as far as I can tell. Essentially, no `kwargs` in the wrapper constructor (as @MiroDudik wrote you can make it a dict or add them individually as I did below, your choice), no model initialization in the wrapper constructor. The model constructor args go on the wrapper constructor explicitly and are passed to the model constructor in the fit method. 

```
class CatBoostClassifierAdapter(BaseEstimator, ClassifierMixin):
    def __init__(self, cat_features=None, text_features=None,
                 iterations=100, learning_rate=0.1, depth=7,
                 loss_function='Logloss', eval_metric="Accuracy",
                 verbose=False, random_state=42):
        self.cat_features = cat_features
        self.text_features = text_features
        self.model = None
        self.iterations = iterations
        self.learning_rate = learning_rate
        self.depth = depth
        self.loss_function = loss_function
        self.eval_metric = eval_metric
        self.verbose = verbose
        self.random_state = random_state

    def fit(self, X, y):
        self.model = CatBoostClassifier(
            iterations=self.iterations,
            learning_rate=self.learning_rate,
            depth=self.depth,
            loss_function=self.loss_function,
            eval_metric=self.eval_metric,
            verbose=self.verbose,
            random_state=self.random_state)
        self.model.fit(
            X, y,
            cat_features=self.cat_features,
            text_features=self.text_features,
            verbose=self.verbose
        )
        return self.model

    def predict(self, X):
        check_is_fitted(self.model)
        return self.model.predict(X)

model1 = CatBoostClassifierAdapter(cat_features=categorical_features)
model2 = CatBoostClassifierAdapter(cat_features=categorical_features)
model3 = clone(estimator=model1, safe=False)
model4 = copy.deepcopy(model1)

model1.fit(X_train, y_train)
model2.fit(X_train, y_train)
model3.fit(X_train, y_train)
model4.fit(X_train, y_train)

y1 = model1.predict(X_test)
y2 = model2.predict(X_test)
y3 = model3.predict(X_test)
y4 = model4.predict(X_test)

print((y1 == y2).all())
print((y1 == y3).all())
print((y1 == y4).all())

# prints True True True
```
and if I look into the wrapper's `model` (e.g., `model1.model.__dict__`) for each of them they are identical, too.

BTW I ran the same with the unchanged wrapper class and the `y3` was different, just as one would expect given what's been explained here so far.

@CarlaFernandez I hope that makes sense. Feel free to respond if you have thoughts, comments, or questions. I'd love to hear if that unblocks you 🙂

**Comment by CarlaFernandez** [comment]
Hello @romanlutz and all, thanks for the support on this issue. 

I originally created the Adapter class simply because I thought I needed my model to inherit from sklearn's BaseEstimator to make it work with fairlearn. Reading the docs more carefully, I now see it just needed to implement `fit` and `predict`. I also didn't notice that `sample_weight` was a valid parameter for CatBoostClassifier already, hence the reassignment inside my `fit` method.

I now think my issue was more due to a lack of attention to detail while reading than anything else! But, I still think it is a really useful thing to document, since people might be using custom models and need to pass extra parameters to the constructor. And of course the fix substituting copy for deepcopy is a nice find!

Thanks again and regards

**Comment by romanlutz** [comment]
It's great feedback, thanks for being so responsive and providing a sample that is easy to reproduce.

@fairlearn/fairlearn-maintainers this is a good reminder to make our requirements of estimators more visible in the docs.

**Comment by adrinjalali** [comment]
I really wouldn't replace `clone` with `deepcopy`.

- `clone` creates a fresh unfitted estimator, whereas `deepcopy` would copy a fitted estimator.
- There is a lot of discussions going on regarding cloning and random state, which we also don't want to deal with here
- users can customize cloning behavior by implementing `__sklearn_clone__`: https://scikit-learn.org/dev/developers/develop.html#cloning

So I'd suggest document well what we expect, how the estimator works, and how people could potentially fix this issue.

**Comment by hildeweerts** [comment]
Thanks @adrinjalali!

i'm not sure whether this was already discussed during one of the calls during my vacation, but @MiroDudik what do you think given adrin's points?

**Comment by immu4989** [comment]
Hi @CarlaFernandez @MiroDudik @hildeweerts @tamaraatanasoska ,I dug into this. The behavior is reproducible on current main, and I think there's a real but bounded fix worth pursuing. Three layer diagnosis below, with a repro test file backing each claim.

Layer 1 : fit() is deterministic. With a deterministic inner estimator (LogisticRegression(random_state=42) or DecisionTreeClassifier(random_state=42)), two calls to ExponentiatedGradient.fit() on identical data produce identical weights_, predictors_, lambda_vecs_EG_, lambda_vecs_LP_, and best_iter_. So the EG iterations themselves are not introducing nondeterminism.

Layer 2 : predict(X, random_state=k) is deterministic. The documented per-call seed works as advertised.

Layer 3 : predict(X) without a seed is nondeterministic when weights_ is a real mixture. This is what @CarlaFernandez almost certainly hit. Concretely:

With eps=0.001, max_iter=50 on a 500-row synthetic dataset, weights_ lands on 3 predictors (~38%/23%/40%), and _pmf_predict returns 7 distinct probability values (not just 0/1).
Two consecutive predict(X) calls disagree on 27/500 rows.
Root cause: predict()'s default random_state=None resolves via check_random_state(None) to numpy's global mtrand._rand. Consecutive calls advance global state and yield different draws inside (positive_probs >= random_state.rand(...)) * 1.
The reason this is easy to miss: when weights_ collapses to a single predictor (common with simple data + few iterations), _pmf_predict outputs are all 0/1 and randomization is a silent no-op — so users who run small examples don't see the issue, but real workloads (CatBoost on census data, more iterations) do.

Repro: [test_reproducibility_1261.py](https://github.com/%3Cimmu4989%3E/fairlearn/blob/issue-1261-investigation/test/unit/reductions/exponentiated_gradient/test_reproducibility_1261.py). Six tests: 4 passing diagnostics that lock in layers 1 and 2 plus the layer 3 failure mode, and 2 xfail(strict=True) tests expressing the proposed fix below.

Proposed fix: add a random_state parameter to ExponentiatedGradient.__init__, used as the default in predict() when no per-call seed is supplied. Sketch:


def __init__(self, estimator, constraints, *, ..., random_state=None):
    ...
    self.random_state = random_state

def predict(self, X, random_state=None):
    ...
    if random_state is None:
        random_state = self.random_state
    random_state = check_random_state(random_state)
    ...
This:

Aligns ExponentiatedGradient with the standard sklearn random_state convention used by every other randomized estimator.
Preserves backwards compatibility: predict(X, random_state=k) still wins, predict(X) with no __init__ seed still uses global state.
Surfaces the randomized-prediction caveat at the constructor — currently it lives only in predict()'s docstring, which users don't read until after they're confused.
Costs ~6 lines of code plus docstring updates.
The same treatment likely makes sense for GridSearch if it has analogous behavior, but I'd scope this PR to ExponentiatedGradient only and address GridSearch separately if relevant.

Two questions before I open a PR:

1) Does this design match how the maintainers want it? Specifically: should random_state on __init__ be stored verbatim and only resolved via check_random_state at predict() call time (so it stays serializable), or eagerly converted to a RandomState instance in __init__? I'd lean toward the former for clone/pickling friendliness.
2) Should I also add a deprecation note pointing users who currently rely on the global state behavior toward the new parameter, or is a CHANGELOG entry sufficient?

Happy to open the PR as soon as I have your steer.

**Comment by immu4989** [comment]
Hi @MiroDudik @hildeweerts @romanlutz @adrinjalali — gentle follow up on the analysis above.

To be explicit on scope: my proposal is independent of the 2023 clone-vs-deepcopy thread. The angle here is purely API consistency at predict-time — ExponentiatedGradient.__init__ doesn't accept random_state, so users following standard sklearn convention have no constructor level way to make the documented prediction randomization reproducible. The fix is a ~6-line addition (store random_state on __init__, use as the default when predict() is called without an explicit seed) and doesn't touch the clone() semantics @adrinjalali raised earlier.

Happy to open the PR scoped narrowly to this if it's acceptable in principle. If you'd rather see it as a separate focused issue to keep this thread on its original topic, I can open one and link back here. A one line preference is all I need.
